test/testindexer.go

```diff
@@ -0,0 +1,200 @@
+		watchers: watchers,
+
+		Ctx: ctx,
+		Dir: dir,
+	}
+
+	// Use a clean environment, with the host's PATH, and a temporary HOME.
+	// We also tell "go install" to place binaries there.
+	hostEnv := os.Environ()
+	var filteredEnv []string
+	for _, env := range hostEnv {
+		if strings.Contains(env, "CC") || strings.Contains(env, "LDFLAGS") || strings.Contains(env, "CFLAGS") {
+			// Bring in the C compiler flags from the host. For example on a Nix
+			// machine, this compilation within the test will fail since the compiler
+			// will not find correct libraries.
+			filteredEnv = append(filteredEnv, env)
+		} else if strings.HasPrefix(env, "PATH") {
+			// Bring in the host's PATH.
+			filteredEnv = append(filteredEnv, env)
+		}
+	}
+	tr.Env = append(filteredEnv, []string{
+		"HOME=" + tr.Dir,
+		"GOBIN=" + tr.Dir,
+	}...)
+	if runtime.GOOS == "windows" {
+		const gopath = "C:\\Projects\\Go"
+		err := os.MkdirAll(gopath, 0666)
+		require.NoError(t, err)
+		tr.Env = append(tr.Env, fmt.Sprintf("GOPATH=%s", gopath))
+	}
+	t.Logf("Env: %s", strings.Join(tr.Env, " "))
+
+	// Reuse the host's build and module download cache.
+	// This should allow "go install" to reuse work.
+	for _, name := range []string{"GOCACHE", "GOMODCACHE"} {
+		out, err := exec.Command("go", "env", name).CombinedOutput()
+		require.NoError(t, err)
+		out = bytes.TrimSpace(out)
+		tr.Env = append(tr.Env, fmt.Sprintf("%s=%s", name, out))
+	}
+
+	return &tr
+}
+
+// Run runs a command and returns its output. This is useful for executing
+// synchronous commands within the temporary environment.
+func (tr *TestIndexerRunner) Run(name string, args ...string) []byte {
+	tr.t.Helper()
+
+	tr.t.Logf("run: %s %s", name, strings.Join(args, " "))
+
+	cmd := exec.CommandContext(tr.Ctx, name, args...)
+	cmd.Env = tr.Env
+	out, err := cmd.CombinedOutput()
+	require.NoError(tr.t, err, string(out))
+	return out
+}
+
+// Start starts a command and returns the command. This is useful for executing
+// asynchronous commands within the temporary environment. It will watch the
+// command's stdout for the given match string, and send on a watcher's
+// channel when/if found.
+func (tr *TestIndexerRunner) Start(prog string, args ...string) *exec.Cmd {
+	tr.t.Helper()
+
+	name := filepath.Base(prog)
+	tr.t.Logf("run: %s %s", name, strings.Join(args, " "))
+
+	cmd := exec.CommandContext(tr.Ctx, prog, args...)
+	cmd.Env = tr.Env
+
+	stdout, err := cmd.StdoutPipe()
+	require.NoError(tr.t, err)
+	cmd.Stderr = cmd.Stdout
+
+	scanner := bufio.NewScanner(stdout)
+
+	go func() {
+		for scanner.Scan() {
+			line := strings.ToLower(scanner.Text())
+
+			// Logging every single line via the test output is verbose,
+			// but helps see what's happening, especially when the test fails.
+			tr.t.Logf("%s: %s", name, line)
+
+			for _, watcher := range tr.watchers {
+				if watcher.Prog == name && strings.Contains(line, strings.ToLower(watcher.Match)) {
+					watcher.Signal <- struct{}{}
+				}
+			}
+		}
+	}()
```

Comment:
I know that is what it did previously, but...

It seems strange to watch every command and see if its output contains something a watcher is looking for.  Would it make more sense if `...watchers` were a parameter to `Start`?

Reply:
A bit tricky because that's two lots of varargs, but if you have a look at the latest commit you'll see my attempt to resolve the problem by introducing a new `Execution` type that lets us build an "execution" and attach watchers to it d2bb69c1